tpmutil/mssim/mssim.go

```diff
@@ -0,0 +1,214 @@
+}
+
+// sendPlatformCommand sends device management commands to the simulator.
+//
+// See: "D.4.3.2. PlatformServer()"
+func sendPlatformCommand(conn net.Conn, buff [4]byte, u uint32) error {
+	binary.BigEndian.PutUint32(buff[:], u)
+	if _, err := conn.Write(buff[:]); err != nil {
+		return fmt.Errorf("write platform command: %v", err)
+	}
+	if _, err := io.ReadFull(conn, buff[:]); err != nil {
+		return fmt.Errorf("read platform command: %v", err)
+	}
+	rc := binary.BigEndian.Uint32(buff[:])
+	if rc != 0 {
+		return fmt.Errorf("unexpected platform command response code: 0x%x", rc)
+	}
+	return nil
+}
+
+const (
+	connStateIdle    = 0
+	connStateRunning = 1
+)
+
+// Conn is a Microsoft Simulator client that can be used as a connection for the
+// tpm2 package.
+type Conn struct {
+	// Cached connection
+	conn net.Conn
+
+	// Response bytes left over from the previous read.
+	prevRead []byte
+}
+
+// Read a response from the simulator. If the response is longer than the provided
+// buffer, the remainer will be cached for the next read.
+func (c *Conn) Read(b []byte) (int, error) {
+	if c.prevRead != nil {
+		n := copy(b, c.prevRead)
+		if len(c.prevRead) >= len(b) {
+			// If lengths are equal return 0 on the next read to indicate there's
+			// nothing left to read.
+			c.prevRead = c.prevRead[n:]
+		} else {
+			c.prevRead = nil
+		}
+		return n, nil
+	}
+
+	// Response frame:
+	// - uint32 (size of response)
+	// - []byte (response)
+	// - uint32 (always 0)
+	var respLen uint32
+	if err := binary.Read(c.conn, binary.BigEndian, &respLen); err != nil {
+		return 0, fmt.Errorf("read MS simulator response header: %v", err)
+	}
+
+	resp := make([]byte, int(respLen)+4)
+	if _, err := io.ReadFull(c.conn, resp[:]); err != nil {
+		return 0, fmt.Errorf("read MS simulator response: %v", err)
+	}
+
+	rc := binary.BigEndian.Uint32(resp[len(resp)-4:])
+	if rc != 0 {
+		return 0, fmt.Errorf("MS simulator returned invalid return code: 0x%x", rc)
+	}
+
+	got := resp[:len(resp)-4]
+
+	n := copy(b, got)
+	if len(got) >= len(b) {
+		c.prevRead = got[n:]
+	}
+	return n, nil
+}
+
+// Write a raw command to the simulator. Commands must be written in a single call
+// to Write. Commands split over multiple calls will result in multiple framed
+// requests.
+func (c *Conn) Write(b []byte) (int, error) {
+	if len(c.prevRead) > 0 {
+		c.prevRead = nil
+	}
+
+	// See: D.4.3.12. TpmServer()
+	buff := &bytes.Buffer{}
+	// "send command" flag
+	if err := binary.Write(buff, binary.BigEndian, tpmSendCommand); err != nil {
+		return 0, err
+	}
+	// locality 0
+	buff.WriteByte(0)
+	// size of the command
+	if err := binary.Write(buff, binary.BigEndian, uint32(len(b))); err != nil {
+		return 0, err
+	}
+	// raw command
+	buff.Write(b)
+
+	if _, err := buff.WriteTo(c.conn); err != nil {
+		return 0, fmt.Errorf("write MS simulator command: %v", err)
+	}
+	return len(b), nil
+}
+
+// Close closes any outgoing connections to the TPM simulator.
+func (c *Conn) Close() error {
```

Comment:
Can you do a graceful shutdown similar to one in Open?

Reply:
done